### PR TITLE
Replace training graph with staged runtime

### PR DESCRIPTION
**Pull request by jwmccarthy**
## Summary
- replace the global training graph with explicit collect, store, sample, prepare, optimize, and runtime stages
- add rollout/replay storage, feed-forward and recurrent PPO, ordered GAIFO stages, model-version validation, and CARL integration
- remove superseded graph, modifier, sampler, update, and circular-buffer code

## Verification
- 19 unit/integration tests
- Ruff checks
- import and compile smoke checks
- one-update CUDA CARL training smoke test

### CODE REVIEW

File: jarl/collect/behavior.py
Comment (jwmccarthy) [inline jarl/collect/behavior.py:1]: I don't know that I like the "behavior" terminology. There's no need to decouple the action from the policy that produces it. Just use the policy architecture we already have?

File: jarl/collect/capture.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            "policy_version": th.full_like(
+                boundary,
+                self.behavior.version,
+                dtype=th.long,
+            )
+        }
+
+
+class ValueCapture:
+    produces = frozenset(
```
Comment (jwmccarthy) [inline jarl/collect/capture.py]: This goes for all "produces" attributes, but do we really need these if we're moving away from the graph? Seems like the only thing these would be good for is validating we have the prerequisite information for each module, but we can just offload that responsibility to the user.

File: jarl/collect/capture.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                    context.env_step.terminated,
+                    device=context.obs.device,
+                ),
+                self.estimator.version,
+                dtype=th.long,
+            ),
+        }
+
+
+def validate_captures(captures: list[Capture] | tuple[Capture, ...]) -> None:
```
Comment (jwmccarthy) [inline jarl/collect/capture.py]: Same thing here: trust the user and avoid automatic validation

File: jarl/collect/capture.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            raise ValueError(f"duplicate capture fields: {sorted(duplicate)}")
+        produced.update(capture.produces)
+
+
+def build_record(
+    context: CaptureContext,
+    captures: list[Capture] | tuple[Capture, ...],
+) -> dict[str, th.Tensor]:
+    record = {
+        "obs": context.obs,
```
Comment (jwmccarthy) [inline jarl/collect/capture.py]: We may as well use observation, action, and reward as the variable names rather than the shortened versions. Keep next_obs though to keep the length similar enough

File: jarl/collect/runner.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        self.state = self.behavior.initial_state(self.n_envs)
+        return self.obs
+
+    @th.no_grad()
+    def step(self) -> EnvStep:
+        if self.obs is None:
+            raise RuntimeError("runner must be reset before stepping")
+
+        obs = th.as_tensor(self.obs, device=self.behavior.device)
+        decision = self.behavior.act(obs, self.state)
```
Comment (jwmccarthy) [inline jarl/collect/runner.py]: Decision is too abstract. action, state/observation, value, policy, etc. are the constructs we should be focusing on because they mirror the math behind RL. No need to abstract the terminology

File: jarl/data/records.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+from dataclasses import dataclass, field
+from typing import Any
+
+import torch as th
+
+
+@dataclass
+class ActionDecision:
+    action: th.Tensor
```
Comment (jwmccarthy) [inline jarl/data/records.py]: Adjust all dataclasses to have the type hints aligned at the least possible column.

File: jarl/envs/gym.py [hunk 6]
```diff
@@ -27,47 +27,38 @@ def __init__(
...
-        return np.stack(obs)
-    
-    def step(self, act: NDArray | th.Tensor = None) -> EnvOutput:
-        actions = act if act is not None else self._sample_acts()
-        if isinstance(act, th.Tensor):
-            actions = act.detach().cpu().numpy()
-        reward, length = [], []   
+        self.terminated = np.empty((n_envs,), dtype=bool)
+        self.truncated = np.empty_like(self.terminated)
+        self.collector_obs = np.empty_like(self.obs)
```
Comment (jwmccarthy) [inline jarl/envs/gym.py]: What is collector_obs? Why not "observation"

File: jarl/learn/gaifo.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        self.section = section
+        self.requires = frozenset({source})
+
+    def run(self, workspace) -> None:
+        rollout = workspace.require(self.source)
+        agent = rollout.steps.select("obs", "next_obs").flatten(0, 1)
+        losses = []
+        for _ in range(self.epochs):
+            indices = th.randperm(len(agent), device=agent.device)
+            for left in range(0, len(agent), self.batch_size):
```
Comment (jwmccarthy) [inline jarl/learn/gaifo.py]: More new lines (here and in general) to delineate semantic blocks and improve readability via visual separation

File: jarl/learn/ppo.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        self.transforms = tuple(transforms)
+        self.optimizer = optimizer
+
+    def update(self, rollout: Rollout) -> dict[str, dict[str, float]]:
+        policy_versions = rollout.steps["policy_version"].unique()
+        if (
+            policy_versions.numel() != 1
+            or policy_versions.item() != self.optimizer.policy.version
+        ):
+            raise RuntimeError("rollout was not collected by the current policy version")
```
Comment (jwmccarthy) [inline jarl/learn/ppo.py]: Just reminding of the semantic newlines for visual separation

File: jarl/learn/program.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    def run(self, workspace: LearningWorkspace) -> None:
+        ...
+
+
+class LearningProgram:
+    def __init__(self, stages) -> None:
+        self.stages = tuple(stages)
+        available = {"source"}
+        for stage in self.stages:
+            missing = stage.requires - available
```
Comment (jwmccarthy) [inline jarl/learn/program.py]: I think I'm on the fence about the "requires" & "produces" infrastructure. It seems like it would be helpful in catching semantic errors, but it is also a bit of a tangle. What do you think?

File: jarl/learn/program.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        steps = apply_transforms(
+            source.steps,
+            self.transforms,
+            PrepareContext(source),
+        )
+        workspace.publish(self.output, source.with_steps(steps))
+
+
+class OptimizePPO:
+    def __init__(self, source: str, optimizer, section: str = "PPO") -> None:
```
Comment (jwmccarthy) [inline jarl/learn/program.py]: What does "source" mean? Same with "sink" earlier. Make the terminology more in line with RL traditional lingo

File: jarl/log/progress.py
Comment (jwmccarthy) [inline jarl/log/progress.py:1]: Let's just abandon custom metric logging altogether. We'll use tensorboard. Keep basic logging for updates & global ticks as a progress indicator in the console

File: jarl/sample/rollout.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class RolloutMinibatches:
+    def __init__(self, batch_size: int, epochs: int = 1) -> None:
+        if batch_size < 1 or epochs < 1:
+            raise ValueError("minibatch settings must be positive")
+        self.batch_size = batch_size
+        self.epochs = epochs
+
+    def __call__(self, data: TensorBatch):
```
Comment (jwmccarthy) [inline jarl/sample/rollout.py:16]: Kind of an ugly nested function

File: jarl/sample/rollout.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        sequences_per_batch: int,
+        epochs: int = 1,
+    ) -> None:
+        if sequence_length < 1 or sequences_per_batch < 1 or epochs < 1:
+            raise ValueError("sequence settings must be positive")
+        self.sequence_length = sequence_length
+        self.sequences_per_batch = sequences_per_batch
+        self.epochs = epochs
+
+    def __call__(self, data: TensorBatch):
```
Comment (jwmccarthy) [inline jarl/sample/rollout.py:60]: Big function here and quite nested. Split into helpers possibly and unnest

File: jarl/store/rollout.py
Comment (jwmccarthy) [inline jarl/store/rollout.py:1]: There must be a way to fuse a lot of the functionality between rollout and replay.

File: jarl/transform/returns.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
+import torch as th
+
+from jarl.data.batch import TensorBatch
+from jarl.transform.base import PrepareContext
+
+
+class GAE:
+    produces = frozenset({"adv", "ret"})
```
Comment (jwmccarthy) [inline jarl/transform/returns.py]: Same thing here with value names. Just use the full ones for clarity

File: jarl/transform/returns.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        self,
+        gamma: float = 0.99,
+        lambda_: float = 0.95,
+        reward_field: str = "rew",
+    ) -> None:
+        self.gamma = gamma
+        self.lambda_ = lambda_
+        self.reward_field = reward_field
+        self.requires = frozenset(
+            {
```
Comment (jwmccarthy) [inline jarl/transform/returns.py]: It's cases like this where the value chain management gets a little verbose for my liking. Maybe we can find a sleeker solution?

File: jarl/transform/returns.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        self.gamma = gamma
+        self.reward_field = reward_field
+        self.requires = frozenset({reward_field, "terminated", "truncated"})
+
+    @th.no_grad()
+    def __call__(
+        self, batch: TensorBatch, context: PrepareContext
+    ) -> TensorBatch:
+        reward = batch[self.reward_field]
+        continuation = ~(batch["terminated"] | batch["truncated"])
```
Comment (jwmccarthy) [inline jarl/transform/returns.py]: Rename to "continue"

File: jarl/collect/behavior.py
Comment (jwmccarthy) [reply jarl/collect/behavior.py:1]: Removed the separate Behavior protocol. Runner now consumes the existing Policy architecture directly, and PolicyOutput is simply the result of policy evaluation.

File: jarl/collect/capture.py (hunk 2 above)
Comment (jwmccarthy) [reply jarl/collect/capture.py]: Removed automatic capture validation and duplicate-output infrastructure. Captures now execute in configured order and update the transition directly.

File: jarl/collect/capture.py (hunk 1 above)
Comment (jwmccarthy) [reply jarl/collect/capture.py]: Removed all produces declarations from captures, transforms, and learning stages. Ordered recipes now trust the user instead of maintaining a parallel field graph.

File: jarl/collect/capture.py (hunk 3 above)
Comment (jwmccarthy) [reply jarl/collect/capture.py]: Applied repo-wide. Canonical fields are now observation, action, reward, next_obs, advantage, and returns.

File: jarl/collect/runner.py (hunk 4 above)
Comment (jwmccarthy) [reply jarl/collect/runner.py]: Renamed ActionDecision to PolicyOutput and removed the artifact layer. It directly exposes action, next_state, and log_prob from Policy.act().

File: jarl/data/records.py (hunk 5 above)
Comment (jwmccarthy) [reply jarl/data/records.py]: Aligned type hints at the least common column across all dataclasses, including records, sequence batches, PPO config, workspace, clock, and Gymnasium tensor-space dataclasses.

File: jarl/envs/gym.py (hunk 6 above)
Comment (jwmccarthy) [reply jarl/envs/gym.py]: Renamed collector_obs to observation. next_obs remains the true transition successor; observation is what the policy receives next after any automatic reset. The distinction is documented on EnvStep.

File: jarl/learn/ppo.py (hunk 8 above)
Comment (jwmccarthy) [reply jarl/learn/ppo.py]: Refactored PPO into sample unpacking, advantage normalization, policy loss, and value loss helpers, with semantic whitespace between evaluation, loss construction, optimization, and metrics.

File: jarl/learn/gaifo.py (hunk 7 above)
Comment (jwmccarthy) [reply jarl/learn/gaifo.py]: Added semantic whitespace throughout the new runtime and specifically separated discriminator sampling, scoring, loss construction, optimization, and metric publication blocks.

File: jarl/learn/program.py (hunk 9 above)
Comment (jwmccarthy) [reply jarl/learn/program.py]: I agree the checks were useful in isolation but recreated too much of the graph we removed. I removed requires/produces/replaces entirely. Stages and transforms run in explicit list order and missing experience fails at its point of use.

File: jarl/learn/program.py (hunk 10 above)
Comment (jwmccarthy) [reply jarl/learn/program.py]: Replaced source/sink throughout the runtime with policy, buffer, experience, rollout, and expert_buffer according to the object’s RL role. TransformArtifact is now TransformRollout.

File: jarl/log/progress.py
Comment (jwmccarthy) [reply jarl/log/progress.py:1]: Removed the custom table, box drawing, ANSI rendering, and colorama dependency. Metrics now go only to TensorBoard; console output is limited to update count and global ticks.

File: jarl/sample/rollout.py (hunk 11 above)
Comment (jwmccarthy) [reply jarl/sample/rollout.py:16]: Split flat sampling into epoch setup and a dedicated _sample_epoch helper. The public call now only validates, flattens, and delegates.

File: jarl/store/rollout.py
Comment (jwmccarthy) [reply jarl/store/rollout.py:1]: Added TensorStorage as the shared backend for rollout and replay. It now owns lazy allocation, environment/shape validation, stable fields, device conversion, and indexed writes; rollout/replay retain only their distinct cursor and lifetime semantics.

File: jarl/sample/rollout.py (hunk 12 above)
Comment (jwmccarthy) [reply jarl/sample/rollout.py:60]: Split recurrent sampling into _build_sequences, _sample_epoch, and _build_batch helpers, reducing nesting and isolating reshaping from sampling and reset-mask construction.

File: jarl/transform/returns.py (hunk 15 above)
Comment (jwmccarthy) [reply jarl/transform/returns.py]: Python reserves continue as a keyword, so I renamed the mask to continues, which is the closest valid direct description.

File: jarl/transform/returns.py (hunk 14 above)
Comment (jwmccarthy) [reply jarl/transform/returns.py]: Removed the value-chain declarations entirely. GAE now contains only its mathematical configuration and calculation; transform ordering is the recipe’s list order.

File: jarl/transform/returns.py (hunk 13 above)
Comment (jwmccarthy) [reply jarl/transform/returns.py]: Expanded these fields and locals throughout the repository: advantage and returns now replace adv and ret, alongside observation/action/reward elsewhere.

File: jarl/learn/program.py (hunk 10 above)
Comment (jwmccarthy) [reply jarl/learn/program.py]: To answer the terminology question directly: source meant the experience object a learner reads from (a finalized rollout or replay buffer), and sink meant the buffer that collection writes into. Those names described dataflow rather than RL concepts, so I removed them: Runner now writes to buffer, Trainer owns buffer, and learning stages name experience, rollout, and expert_buffer explicitly.
